File: Orchestrator/src/orchestrator/nodeController.js
import { EventEmitter } from "events";
import { RS485Comm } from "./rs485Comm.js";

/**
 * Node States
 */
export const NodeState = {
  IDLE: "IDLE",
  CONFIGURED: "CONFIGURED",
  RUNNING: "RUNNING",
  STOPPED: "STOPPED",
  ERROR: "ERROR",
};

/**
 * Node Controller
 * High-level control for UWB nodes (Node A = TX, Node B = RX)
 */
export class NodeController extends EventEmitter {
  constructor(nodeId, rs485Comm) {
    super();
    this.nodeId = nodeId; // 'A' or 'B'
    this.rs485Comm = rs485Comm;
    this.state = NodeState.IDLE;
    this.config = null;
    this.stats = null;
    this.lastError = null;

    // Bind RS-485 events
    this.rs485Comm.on("error", (error) => {
      this.setState(NodeState.ERROR);
      this.lastError = error.message;
      // Only emit error if there are listeners to prevent uncaught exceptions
      if (this.listenerCount("error") > 0) {
        this.emit("error", error);
      } else {
        // Log error if no listeners (prevents uncaught exception)
        console.error(
          `[NodeController ${this.nodeId}] Unhandled error:`,
          error.message,
        );
      }
    });

    this.rs485Comm.on("open", () => {
      this.emit("connected");
    });

    this.rs485Comm.on("close", () => {
      this.setState(NodeState.IDLE);
      this.emit("disconnected");
    });
  }

  /**
   * Set node state
   * @param {string} newState
   */
  setState(newState) {
    if (this.state !== newState) {
      const oldState = this.state;
      this.state = newState;
      this.emit("stateChange", { oldState, newState, nodeId: this.nodeId });
    }
  }

  /**
   * Get current state
   * @returns {string}
   */
  getState() {
    return this.state;
  }

  /**
   * Check if node is connected
   * @returns {boolean}
   */
  isConnected() {
    return this.rs485Comm.isOpen;
  }

  /**
   * Open connection to node
   */
  async connect() {
    try {
      await this.rs485Comm.open();
      return true;
    } catch (error) {
      this.setState(NodeState.ERROR);
      this.lastError = error.message;
      throw error;
    }
  }

  /**
   * Close connection to node
   */
  async disconnect() {
    try {
      await this.rs485Comm.close();
      this.setState(NodeState.IDLE);
      return true;
    } catch (error) {
      this.lastError = error.message;
      throw error;
    }
  }

  /**
   * Ping node to check connectivity
   * @returns {Promise<boolean>}
   */
  async ping() {
    try {
      // Use longer timeout for Node B (which is less reliable)
      const timeout = this.nodeId === "B" ? 8000 : 5000;

      const response = await this.rs485Comm.sendCommand("PNG", timeout);

      return response.startsWith("OK");
    } catch (error) {
      this.lastError = error.message;
      console.error(
        `⚠️  ${this.nodeId} PING failed - firmware may not be responding`,
      );
      console.error(`   Error: ${error.message.split("\n")[0]}`);
      if (error.message.includes("timeout")) {
        console.error(`   No response received. Check:`);
        console.error(`   1. Firmware is running orchestrator example`);
        console.error(`   2. RS-485 hardware connection`);
        console.error(`   3. Serial port ${this.rs485Comm.port} is correct`);
        console.error(
          `   4. Baud rate: Should be 115200 for RS485 communication (orchestrator to Arduino bridge)`,
        );
        console.error(
          `   5. If you see corrupted data (high-bit bytes), check baud rate mismatch`,
        );
        if (this.nodeId === "B") {
          console.error(
            `   6. Node B is less reliable - check RS-485 wiring and termination`,
          );
        }
      }
      return false;
    }
  }

  /**
   * Get firmware node type
   * @returns {Promise<string|null>}
   */
  async getFirmwareNodeType() {
    try {
      // Node B needs longer timeout
      const timeout = this.nodeId === "B" ? 8000 : 5000;
      const response = await this.rs485Comm.getNodeType(timeout);
      if (response && response.startsWith("OK NODE_TYPE=")) {
        const type = response.split("=")[1];
        const trimmedType = type.trim();
        return trimmedType;
      }
      return null;
    } catch (error) {
      this.lastError = error.message;
      return null;
    }
  }

  /**
   * Configure node with UWB settings
   * @param {Object} config - UWB configuration
   * @returns {Promise<boolean>}
   */
  async configure(config) {
    try {
      console.log(
        `[NodeController ${this.nodeId}] configure() called - state: ${this.state}, connected: ${this.isConnected()}`,
      );

      if (this.state === NodeState.RUNNING) {
        throw new Error("Cannot configure while test is running");
      }

      // Try to open connection if not already open
      if (!this.isConnected()) {
        console.log(
          `[NodeController ${this.nodeId}] Port not open, attempting to open...`,
        );
        try {
          await this.connect();
        } catch (connectError) {
          console.warn(
            `[NodeController ${this.nodeId}] Failed to open connection: ${connectError.message}, continuing anyway...`,
          );
        }
      }

      console.log(`[NodeController ${this.nodeId}] Sending configuration...`);
      // Configuration can take time - dwt_configure() and configure_tx_power() may take several seconds
      // Use longer timeout for configuration commands (20 seconds should be enough)
      const timeout = this.nodeId === "B" ? 20000 : 20000;
      const response = await this.rs485Comm.setConfig(config, timeout);
      console.log(
        `[NodeController ${this.nodeId}] Configuration response: ${response}`,
      );

      // Firmware responds with "OK CONFIG" (not "OK CONFIG_SET")
      if (response.startsWith("OK CONFIG")) {
        this.config = { ...config };
        this.setState(NodeState.CONFIGURED);
        this.emit("configured", config);
        const greenColor = "\x1b[32m";
        const resetColor = "\x1b[0m";
        console.log(
          `${greenColor}[NodeController ${this.nodeId}] ✓ Configuration successful${resetColor}`,
        );
        return true;
      } else {
        const errorMsg = `Configuration failed: ${response}`;
        console.error(`[NodeController ${this.nodeId}] ${errorMsg}`);
        throw new Error(errorMsg);
      }
    } catch (error) {
      console.error(
        `[NodeController ${this.nodeId}] configure() error: ${error.message}`,
      );
      this.setState(NodeState.ERROR);
      this.lastError = error.message;
      this.emit("error", error);
      throw error;
    }
  }

  /**
   * Start test
   * @returns {Promise<boolean>}
   */
  async startTest() {
    try {
      console.log(
        `[NodeController ${this.nodeId}] startTest() called - state: ${this.state}, connected: ${this.isConnected()}`,
      );

      // Try to open connection if not already open
      if (!this.isConnected()) {
        console.log(
          `[NodeController ${this.nodeId}] Port not open, attempting to open...`,
        );
        try {
          await this.connect();
        } catch (connectError) {
          console.warn(
            `[NodeController ${this.nodeId}] Failed to open connection: ${connectError.message}, continuing anyway...`,
          );
        }
      }

      console.log(`[NodeController ${this.nodeId}] Sending START command...`);
      // Node B needs longer timeout
      const timeout = this.nodeId === "B" ? 8000 : 5000;
      const response = await this.rs485Comm.startTest(timeout);
      console.log(
        `[NodeController ${this.nodeId}] START response: ${response}`,
      );

      // Firmware responds with "OK START" (not "OK TEST_STARTED")
      if (response.startsWith("OK START")) {
        this.setState(NodeState.RUNNING);
        this.emit("testStarted");
        const greenColor = "\x1b[32m";
        const resetColor = "\x1b[0m";
        console.log(
          `${greenColor}[NodeController ${this.nodeId}] ✓ Test started successfully${resetColor}`,
        );
        return true;
      } else {
        const errorMsg = `Start test failed: ${response}`;
        console.error(`[NodeController ${this.nodeId}] ${errorMsg}`);
        throw new Error(errorMsg);
      }
    } catch (error) {
      console.error(
        `[NodeController ${this.nodeId}] startTest() error: ${error.message}`,
      );
      this.setState(NodeState.ERROR);
      this.lastError = error.message;
      this.emit("error", error);
      throw error;
    }
  }

  /**
   * Stop test
   * @returns {Promise<boolean>}
   */
  async stopTest() {
    try {
      // Try to open connection if not already open
      if (!this.isConnected()) {
        console.log(
          `[NodeController ${this.nodeId}] Port not open, attempting to open...`,
        );
        try {
          await this.connect();
        } catch (connectError) {
          console.warn(
            `[NodeController ${this.nodeId}] Failed to open connection: ${connectError.message}, continuing anyway...`,
          );
        }
      }

      // Node B needs longer timeout
      const timeout = this.nodeId === "B" ? 8000 : 5000;
      const response = await this.rs485Comm.stopTest(timeout);
      if (response.startsWith("OK STOP")) {
        this.setState(NodeState.STOPPED);
        this.emit("testStopped");
        return true;
      } else {
        throw new Error(`Stop test failed: ${response}`);
      }
    } catch (error) {
      this.setState(NodeState.ERROR);
      this.lastError = error.message;
      this.emit("error", error);
      throw error;
    }
  }

  /**
   * Get statistics from node
   * @returns {Promise<Object>}
   */
  async getStats() {
    try {
      // Try to open connection if not already open
      if (!this.isConnected()) {
        console.log(
          `[NodeController ${this.nodeId}] Port not open, attempting to open...`,
        );
        try {
          await this.connect();
        } catch (connectError) {
          console.warn(
            `[NodeController ${this.nodeId}] Failed to open connection: ${connectError.message}, continuing anyway...`,
          );
        }
      }

      // Node B needs longer timeout
      const timeout = this.nodeId === "B" ? 8000 : 5000;
      const response = await this.rs485Comm.getStats(timeout);
      if (response.startsWith("OK STATS")) {
        this.stats = this.parseStats(response);
        this.emit("statsUpdated", this.stats);
        return this.stats;
      } else {
        throw new Error(`Get stats failed: ${response}`);
      }
    } catch (error) {
      this.lastError = error.message;
      throw error;
    }
  }

  /**
   * Parse stats from GET_STATS response
   * Format (Node B): OK STATS total_rx=100 lost_pkts=5 crc_err=2 rssi_avg=-65 snr_avg=25 pre_q_avg=128
   * Format (Node A): OK STATS total_sent=100 last_error=0
   * @param {string} response
   * @returns {Object}
   */
  parseStats(response) {
    const stats = {};
    const parts = response.split(" ");

    for (let i = 2; i < parts.length; i++) {
      const [key, value] = parts[i].split("=");
      if (key && value !== undefined) {
        const numValue = parseFloat(value);
        stats[key] = isNaN(numValue) ? value : numValue;
      }
    }

    // Map firmware v2 format to expected format
    // Firmware sends: sent=, attempted=, errors=, timeouts=, last_err=, frame_dur=
    // But parser expects: total_sent, total_attempted, tx_errors, tx_timeouts, last_error, frame_duration_us
    if (stats.sent !== undefined && stats.total_sent === undefined) {
      stats.total_sent = stats.sent;
    }
    if (stats.attempted !== undefined && stats.total_attempted === undefined) {
      stats.total_attempted = stats.attempted;
    }
    if (stats.errors !== undefined && stats.tx_errors === undefined) {
      stats.tx_errors = stats.errors;
    }
    if (stats.timeouts !== undefined && stats.tx_timeouts === undefined) {
      stats.tx_timeouts = stats.timeouts;
    }
    if (stats.last_err !== undefined && stats.last_error === undefined) {
      stats.last_error = stats.last_err;
    }
    if (
      stats.frame_dur !== undefined &&
      stats.frame_duration_us === undefined
    ) {
      stats.frame_duration_us = stats.frame_dur;
    }

    // Node A (TX) stats
    if (this.nodeId === "A") {
      // Firmware v2 sends: sent=, attempted=, errors=, timeouts=, last_err=, frame_dur=
      // Map firmware format to expected format
      const result = {
        total_sent: stats.total_sent || stats.sent || 0,
        total_attempted: stats.total_attempted || stats.attempted || 0,
        tx_errors: stats.tx_errors || stats.errors || 0,
        tx_timeouts: stats.tx_timeouts || stats.timeouts || 0,
        last_error: stats.last_error || stats.last_err || 0,
        frame_duration_us: stats.frame_duration_us || stats.frame_dur || 0,
      };
      return result;
    }

    // Node B (RX) stats
    // Handle case where Node B firmware might be built as Node A type
    // If we get total_sent instead of total_rx, it means wrong firmware type
    if (stats.total_sent !== undefined && stats.total_rx === undefined) {
      console.warn(
        `[NodeController] Node B returned total_sent instead of total_rx - firmware may be built as Node A type!`,
      );
      console.warn(`[NodeController] Response was: ${response}`);
      // Return zeros for RX stats since we can't parse TX stats as RX stats
      return {
        total_rx: 0,
        lost_pkts: 0,
        crc_err: 0,
        rssi_avg_dbm: 0,
        snr_avg_db: 0,
        preamble_q_avg: 0,
      };
    }

    // Normal Node B (RX) stats
    // Firmware v2 sends: rx=, lost=, crc_err=, phy_err=, timeout=, overrun=, rssi_min=, rssi_max=, pre_q_min=, pre_q_max=
    // Map to expected format
    const result = {
      total_rx: stats.total_rx || stats.rx || 0,
      lost_pkts: stats.lost_pkts || stats.lost || 0,
      crc_err: stats.crc_err || 0,
      phy_err: stats.phy_err || 0,
      rx_timeout: stats.rx_timeout || stats.timeout || 0,
      rx_overrun: stats.rx_overrun || stats.overrun || 0,
      rssi_avg_dbm: stats.rssi_avg || 0, // Calculated from rssi_min/rssi_max if needed
      rssi_min: stats.rssi_min || 0,
      rssi_max: stats.rssi_max || 0,
      snr_avg_db: stats.snr_avg || 0,
      preamble_q_avg: stats.pre_q_avg || 0,
      pre_q_min: stats.pre_q_min || 0,
      pre_q_max: stats.pre_q_max || 0,
    };
    return result;
  }

  /**
   * Reset statistics
   * @returns {Promise<boolean>}
   */
  async resetStats() {
    try {
      // Node B needs longer timeout
      const timeout = this.nodeId === "B" ? 8000 : 5000;
      const response = await this.rs485Comm.resetStats(timeout);
      if (response.startsWith("OK")) {
        this.stats = null;
        this.emit("statsReset");
        return true;
      } else {
        throw new Error(`Reset stats failed: ${response}`);
      }
    } catch (error) {
      this.lastError = error.message;
      throw error;
    }
  }

  /**
   * Get current configuration
   * @returns {Object|null}
   */
  getConfig() {
    return this.config;
  }

  /**
   * Get current statistics
   * @returns {Object|null}
   */
  getStatsSync() {
    return this.stats;
  }

  /**
   * Get last error
   * @returns {string|null}
   */
  getLastError() {
    return this.lastError;
  }
}

File: Orchestrator/python_backend/src/orchestrator/node_controller.py
"""
Node Controller
High-level control for UWB nodes (Node A = TX, Node B = RX)
"""

import asyncio
from enum import Enum
from typing import Optional, Dict, Any
from .rs485_comm import RS485Comm


class NodeState(str, Enum):
    """Node States"""
    IDLE = "IDLE"
    CONFIGURED = "CONFIGURED"
    RUNNING = "RUNNING"
    STOPPED = "STOPPED"
    ERROR = "ERROR"


class NodeController:
    """Node Controller for UWB nodes"""
    
    def __init__(self, node_id: str, rs485_comm: RS485Comm):
        self.node_id = node_id  # 'A' or 'B'
        self.rs485_comm = rs485_comm
        self.state = NodeState.IDLE
        self.config: Optional[Dict[str, Any]] = None
        self.stats: Optional[Dict[str, Any]] = None
        self.last_error: Optional[str] = None
        self._state_callbacks = []
        self._error_callbacks = []
    
    def set_state(self, new_state: NodeState) -> None:
        """Set node state"""
        if self.state != new_state:
            old_state = self.state
            self.state = new_state
            
            # Notify callbacks
            for callback in self._state_callbacks:
                try:
                    callback(old_state, new_state, self.node_id)
                except Exception as e:
                    print(f"[NodeController {self.node_id}] State callback error: {e}")
    
    def get_state(self) -> NodeState:
        """Get current state"""
        return self.state
    
    def is_connected(self) -> bool:
        """Check if node is connected"""
        return self.rs485_comm.is_open
    
    async def connect(self) -> bool:
        """Open connection to node"""
        try:
            await self.rs485_comm.open()
            return True
        except Exception as error:
            self.set_state(NodeState.ERROR)
            self.last_error = str(error)
            raise
    
    async def disconnect(self) -> bool:
        """Close connection to node"""
        try:
            await self.rs485_comm.close()
            self.set_state(NodeState.IDLE)
            return True
        except Exception as error:
            self.last_error = str(error)
            raise
    
    async def ping(self) -> bool:
        """Ping node to check connectivity"""
        try:
            # Use longer timeout for initial PING
            response = await self.rs485_comm.send_command("PNG", 3000)
            return response.startswith("OK")
        except Exception as error:
            self.last_error = str(error)
            print(f"⚠️  {self.node_id} PING failed - firmware may not be responding")
            print(f"   Error: {str(error).split(chr(10))[0]}")
            if 'timeout' in str(error).lower():
                print(f"   No response received. Check:")
                print(f"   1. Firmware is running orchestrator example")
                print(f"   2. RS-485 hardware connection")
                print(f"   3. Serial port {self.rs485_comm.port} is correct")
            return False
    
    async def get_firmware_node_type(self) -> Optional[str]:
        """Get firmware node type"""
        try:
            response = await self.rs485_comm.get_node_type()
            if response and response.startswith("OK NODE_TYPE="):
                node_type = response.split("=")[1].strip()
                return node_type
            return None
        except Exception as error:
            self.last_error = str(error)
            return None
    
    async def configure(self, config: Dict[str, Any]) -> bool:
        """Configure node with UWB settings"""
        try:
            print(f"[NodeController {self.node_id}] configure() called - state: {self.state}, connected: {self.is_connected()}")
            
            if self.state == NodeState.RUNNING:
                raise Exception("Cannot configure while test is running")
            
            if not self.is_connected():
                error_msg = f"Node {self.node_id} is not connected. Call connect() first."
                print(f"[NodeController {self.node_id}] {error_msg}")
                raise Exception(error_msg)
            
            # Quick connectivity check - send PING to verify firmware is responding
            print(f"[NodeController {self.node_id}] Sending PING to verify connectivity...")
            try:
                await self.rs485_comm.ping()
                print(f"[NodeController {self.node_id}] ✓ PING successful")
            except Exception as ping_error:
                error_msg = (
                    f"Node {self.node_id} is not responding to commands. "
                    f"Port is open but firmware may not be running or RS-485 communication is failing. "
                    f"Original error: {ping_error}"
                )
                print(f"[NodeController {self.node_id}] {error_msg}")
                raise Exception(error_msg)
            
            print(f"[NodeController {self.node_id}] Sending configuration...")
            response = await self.rs485_comm.set_config(config)
            print(f"[NodeController {self.node_id}] Configuration response: {response}")
            
            if response.startswith("OK CONFIG"):
                self.config = config.copy()
                self.set_state(NodeState.CONFIGURED)
                print(f"[NodeController {self.node_id}] ✓ Configuration successful")
                return True
            else:
                error_msg = f"Configuration failed: {response}"
                print(f"[NodeController {self.node_id}] {error_msg}")
                raise Exception(error_msg)
                
        except Exception as error:
            print(f"[NodeController {self.node_id}] configure() error: {error}")
            self.set_state(NodeState.ERROR)
            self.last_error = str(error)
            for callback in self._error_callbacks:
                try:
                    callback(error)
                except:
                    pass
            raise
    
    async def start_test(self) -> bool:
        """Start test"""
        try:
            print(f"[NodeController {self.node_id}] startTest() called - state: {self.state}, connected: {self.is_connected()}")
            
            if self.state not in [NodeState.CONFIGURED, NodeState.STOPPED]:
                error_msg = f"Cannot start test from state: {self.state} (must be CONFIGURED or STOPPED)"
                print(f"[NodeController {self.node_id}] {error_msg}")
                raise Exception(error_msg)
            
            if not self.is_connected():
                error_msg = f"Node {self.node_id} is not connected"
                print(f"[NodeController {self.node_id}] {error_msg}")
                raise Exception(error_msg)
            
            print(f"[NodeController {self.node_id}] Sending START command...")
            response = await self.rs485_comm.start_test()
            print(f"[NodeController {self.node_id}] START response: {response}")
            
            if response.startswith("OK START"):
                self.set_state(NodeState.RUNNING)
                print(f"[NodeController {self.node_id}] ✓ Test started successfully")
                return True
            else:
                error_msg = f"Start test failed: {response}"
                print(f"[NodeController {self.node_id}] {error_msg}")
                raise Exception(error_msg)
                
        except Exception as error:
            print(f"[NodeController {self.node_id}] startTest() error: {error}")
            self.set_state(NodeState.ERROR)
            self.last_error = str(error)
            for callback in self._error_callbacks:
                try:
                    callback(error)
                except:
                    pass
            raise
    
    async def stop_test(self) -> bool:
        """Stop test"""
        try:
            if self.state != NodeState.RUNNING:
                # Allow stopping from any state
                return True
            
            response = await self.rs485_comm.stop_test()
            if response.startswith("OK STOP"):
                self.set_state(NodeState.STOPPED)
                return True
            else:
                raise Exception(f"Stop test failed: {response}")
                
        except Exception as error:
            self.set_state(NodeState.ERROR)
            self.last_error = str(error)
            for callback in self._error_callbacks:
                try:
                    callback(error)
                except:
                    pass
            raise
    
    async def get_stats(self) -> Dict[str, Any]:
        """Get statistics from node"""
        try:
            response = await self.rs485_comm.get_stats()
            if response.startswith("OK STATS"):
                self.stats = self._parse_stats(response)
                return self.stats
            else:
                raise Exception(f"Get stats failed: {response}")
        except Exception as error:
            self.last_error = str(error)
            raise
    
    def _parse_stats(self, response: str) -> Dict[str, Any]:
        """
        Parse stats from GET_STATS response
        Format (Node B): OK STATS total_rx=100 lost_pkts=5 crc_err=2 rssi_avg=-65 snr_avg=25 pre_q_avg=128
        Format (Node A): OK STATS total_sent=100 last_error=0
        """
        stats = {}
        parts = response.split(" ")
        
        for part in parts[2:]:  # Skip "OK STATS"
            if "=" in part:
                key, value = part.split("=", 1)
                try:
                    # Try to convert to number
                    num_value = float(value)
                    stats[key] = int(num_value) if num_value.is_integer() else num_value
                except ValueError:
                    stats[key] = value
        
        # Node A (TX) stats
        if self.node_id == "A":
            return {
                'total_sent': stats.get('total_sent', stats.get('sent', 0)),
                'last_error': stats.get('last_error', 0),
            }
        
        # Node B (RX) stats
        # Handle case where Node B firmware might be built as Node A type
        if 'total_sent' in stats and 'total_rx' not in stats:
            print(f"[NodeController] Node B returned total_sent instead of total_rx - firmware may be built as Node A type!")
            print(f"[NodeController] Response was: {response}")
            return {
                'total_rx': 0,
                'lost_pkts': 0,
                'crc_err': 0,
                'rssi_avg_dbm': 0,
                'snr_avg_db': 0,
                'preamble_q_avg': 0,
            }
        
        return {
            'total_rx': stats.get('total_rx', stats.get('rx', 0)),
            'lost_pkts': stats.get('lost_pkts', stats.get('lost', 0)),
            'crc_err': stats.get('crc_err', 0),
            'rssi_avg_dbm': stats.get('rssi_avg', 0),
            'snr_avg_db': stats.get('snr_avg', 0),
            'preamble_q_avg': stats.get('pre_q_avg', 0),
        }
    
    async def reset_stats(self) -> bool:
        """Reset statistics"""
        try:
            response = await self.rs485_comm.reset_stats()
            if response.startswith("OK"):
                self.stats = None
                return True
            else:
                raise Exception(f"Reset stats failed: {response}")
        except Exception as error:
            self.last_error = str(error)
            raise
    
    def get_config(self) -> Optional[Dict[str, Any]]:
        """Get current configuration"""
        return self.config
    
    def get_stats_sync(self) -> Optional[Dict[str, Any]]:
        """Get current statistics (synchronous)"""
        return self.stats
    
    def get_last_error(self) -> Optional[str]:
        """Get last error"""
        return self.last_error
    
    def on_state_change(self, callback) -> None:
        """Register state change callback"""
        self._state_callbacks.append(callback)
    
    def on_error(self, callback) -> None:
        """Register error callback"""
        self._error_callbacks.append(callback)
